### PR TITLE
HPCC-25396 K8s dfuserver: support for spraying

### PR DESCRIPTION
**Pull request by shamser**
Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

### CODE REVIEW

File: dali/dfu/dfurun.cpp [hunk 1]
```diff
@@ -1314,6 +1314,16 @@ class CDFUengine: public CInterface, implements IDFUengine
                                 };
                             }
                         }
+#ifdef _CONTAINERIZED
+                        StringBuffer clusterName;
+                        destination->getGroupName(0, clusterName);
+                        Owned<IPropertyTree> plane = getDropZonePlane(clusterName);
+                        if (plane)
+                        {
+                            unsigned np = plane->getPropInt("@defaultSprayParts", 0);
```
Comment (jakesmith) [inline dali/dfu/dfurun.cpp:1323]: minor: If not set, then 0, which is default internally for numpartsoverride, which means unset.

Would perhaps be clearer intention, if didn't call if not set, e.g.:
```
    if (plane->hasProp("@defaultSprayParts")
        destination->setNumPartsOverride(plane->getPropInt("@defaultSprayParts"));
```

File: esp/services/ws_fs/ws_fsService.cpp [hunk 2]
```diff
@@ -1870,13 +1870,32 @@ bool CFileSprayEx::onSprayFixed(IEspContext &context, IEspSprayFixed &req, IEspS
...
+#else
+        bool isValidDfuQueueName = false;
+        Owned<IPropertyTreeIterator> dfuServers = queryComponentConfig().getElements("dfuQueues");
+        ForEach(*dfuServers)
+        {
+            IPropertyTree & dfuServer = dfuServers->query();
+            const char * dfuServerName = dfuServer.queryProp("@name");
+            StringBuffer knownDfuQueueName;
+            getDfuQueueName(knownDfuQueueName, dfuServerName);
+            if (strcmp(dfuQueue, knownDfuQueueName)==0)
```
Comment (jakesmith) [inline esp/services/ws_fs/ws_fsService.cpp:1890]: trivial: could use streq

File: esp/services/ws_fs/ws_fsService.cpp [hunk 3]
```diff
@@ -3434,7 +3453,17 @@ bool CFileSprayEx::onGetSprayTargets(IEspContext &context, IEspGetSprayTargetsRe
...
         context.ensureFeatureAccess(FILE_SPRAY_URL, SecAccess_Read, ECLWATCH_FILE_SPRAY_ACCESS_DENIED, "Permission denied.");
-
+#ifdef _CONTAINERIZED
+        IArrayOf<IEspGroupNode> sprayTargets;
+        Owned<IPropertyTreeIterator> dataPlanes = queryGlobalConfig().getElements("storage/planes[labels='data']");
+        ForEach(*dataPlanes)
+        {
+            IPropertyTree & plane = dataPlanes->query();
+            const char * name = plane.queryProp("@name");
+            appendGroupNode(sprayTargets, name, name, false /*replicate outputs*/);
```
Comment (jakesmith) [inline esp/services/ws_fs/ws_fsService.cpp:3463]: is it correct to pass 'name' to 3rd param (clusterType)  ?
I guess it's a bit meaningless in cloud, it's usually "hthor", "thor", or "roxie" ..
Is it just for display purposes by the client? Perhaps should just be a generic "storage plane"

File: esp/smc/SMCLib/TpWrapper.cpp [hunk 4]
```diff
@@ -464,8 +464,52 @@ void CTpWrapper::getTpEspServers(IArrayOf<IConstTpEspServer>& list)
     }
 }
 
+static Owned<IEspTpMachine> createLocalHostTpMachine(const char *path)
```
Comment (jakesmith) [inline esp/smc/SMCLib/TpWrapper.cpp:467]: non-standard way of dealing with owner transfer.
should return a regular pointer to IEspTpMachine, caller then owns.

File: esp/smc/SMCLib/TpWrapper.cpp [hunk 5]
```diff
@@ -464,8 +464,52 @@ void CTpWrapper::getTpEspServers(IArrayOf<IConstTpEspServer>& list)
...
+        const char * dfuName = dfuQueue.queryProp("@name");
+        StringBuffer queue;
+        getDfuQueueName(queue, dfuName);
+        Owned<IEspTpDfuServer> pService = createTpDfuServer("","");
+        pService->setName(dfuName);
+        pService->setDescription(dfuName);
+        pService->setBuild("");
+        pService->setQueue(queue);
+        pService->setType(eqDfu);
+        pService->setLogDirectory("/var/log/HPCCSystems/dfuserver");
```
Comment (jakesmith) [inline esp/smc/SMCLib/TpWrapper.cpp:498]: is this needed for now for some backward compact. reason?
there is no log directory in cloud.

File: esp/smc/SMCLib/TpWrapper.cpp [hunk 6]
```diff
@@ -464,8 +464,52 @@ void CTpWrapper::getTpEspServers(IArrayOf<IConstTpEspServer>& list)
...
+        IArrayOf<IEspTpMachine> tpMachines;
+        Owned<IPropertyTreeIterator> planes = getDropZonePlanesIterator();
+        ForEach(*planes)
+        {
+            IPropertyTree & plane = planes->query();
+            const char * path = plane.queryProp("@prefix");
+
+            if (!isEmptyString(path))
+                tpMachines.append(*createLocalHostTpMachine(path).getClear());
```
Comment (jakesmith) [inline esp/smc/SMCLib/TpWrapper.cpp:507]: createLocalHostTpMachine should just return a regular pointer that is owned by caller semantically

File: esp/smc/SMCLib/TpWrapper.cpp [hunk 7]
```diff
@@ -464,8 +464,52 @@ void CTpWrapper::getTpEspServers(IArrayOf<IConstTpEspServer>& list)
...
+        }
+        pService->setTpMachines(tpMachines);
+        list.append(*pService.getLink());
```
Comment (jakesmith) [inline esp/smc/SMCLib/TpWrapper.cpp:510]: trivial: could be getClear here - slightly more efficient.

File: esp/smc/SMCLib/TpWrapper.cpp [hunk 8]
```diff
@@ -1681,27 +1726,21 @@ void CTpWrapper::getTpDropZones(double clientVersion, const char* name, bool ECL
...
+            dropZone->setName(dropzonename);
+            dropZone->setDescription(""); //TODO
+            dropZone->setPath(path);
+            dropZone->setBuild("");
+            dropZone->setECLWatchVisible(plane.getPropBool("@eclWatchVisible", true));
+            StringBuffer localHost;
+            IArrayOf<IEspTpMachine> tpMachines;
+            tpMachines.append(*createLocalHostTpMachine(path).getClear());
+            dropZone->setTpMachines(tpMachines);
+            list.append(*dropZone.getLink());
```
Comment (jakesmith) [inline esp/smc/SMCLib/TpWrapper.cpp:1742]: trivial: could be getClear here - slightly more efficient.

File: esp/smc/SMCLib/TpWrapper.cpp [hunk 9]
```diff
@@ -1681,27 +1726,21 @@ void CTpWrapper::getTpDropZones(double clientVersion, const char* name, bool ECL
...
-        ipAddr.getIpText(localHost);
-        machine->setNetaddress(localHost.str());
-        machine->setConfigNetaddress("localhost");
-        machine->setDirectory(path);
-        machine->setOS(getPathSepChar(path) == '/' ? MachineOsLinux : MachineOsW2K);
-        tpMachines.append(*machine.getLink());
-        dropZone->setTpMachines(tpMachines);
-        list.append(*dropZone.getLink());
+        const char * path = plane.queryProp("@prefix");
+        if (!isEmptyString(path))
```
Comment (jakesmith) [inline esp/smc/SMCLib/TpWrapper.cpp:1730]: minor: should never be possible, schema should reject it not set

File: esp/smc/SMCLib/TpWrapper.cpp [hunk 10]
```diff
@@ -1681,27 +1726,21 @@ void CTpWrapper::getTpDropZones(double clientVersion, const char* name, bool ECL
...
+        {
+            Owned<IEspTpDropZone> dropZone = createTpDropZone();
+            dropZone->setName(dropzonename);
+            dropZone->setDescription(""); //TODO
+            dropZone->setPath(path);
+            dropZone->setBuild("");
+            dropZone->setECLWatchVisible(plane.getPropBool("@eclWatchVisible", true));
```
Comment (jakesmith) [inline esp/smc/SMCLib/TpWrapper.cpp:1737]: eclWatchVisible isn't part of the schema for a plane - doubt we want it to be.
I think for now, should always set to true here.

File: esp/smc/SMCLib/TpWrapper.cpp [hunk 11]
```diff
@@ -1681,27 +1726,21 @@ void CTpWrapper::getTpDropZones(double clientVersion, const char* name, bool ECL
...
+            StringBuffer localHost;
+            IArrayOf<IEspTpMachine> tpMachines;
+            tpMachines.append(*createLocalHostTpMachine(path).getClear());
```
Comment (jakesmith) [inline esp/smc/SMCLib/TpWrapper.cpp:1740]: createLocalHostTpMachine should transfer ownership / return regular pointer

File: helm/hpcc/values.yaml [hunk 12]
```diff
@@ -95,6 +95,7 @@ storage:
   #   includeDeviceInPath: false            # Is the device number appended to the mount for the physical path for a file?  (Only required in unusual situations)
   #   hosts: <name>                         # Name of the host group for bare metal - must match the name of the storage plane..
   #   secret: <secret-id>                   # what secret is required to access the files.  This could optionally become a list if required (or add secrets:).
+  #   defaultSprayParts: 4                  # the partitions created when spraying (if not explicity specified)
```
Comment (jakesmith) [inline helm/hpcc/values.yaml:98]: what happens if not set, and the plane is spray to, without explicitly saying how many parts?
Does it fallback to some global default somewhere, or it an error?

**Review by jakesmith — COMMENTED** [review]
@shamser - looks good, just a few minor comments/questions.

File: esp/smc/SMCLib/TpWrapper.cpp (hunk 5 above)
Comment (shamser) [reply esp/smc/SMCLib/TpWrapper.cpp:498]: I don't see an ill-effects from not setting this.

File: helm/hpcc/values.yaml (hunk 12 above)
Comment (shamser) [reply helm/hpcc/values.yaml:98]: If this is not set, it defaults to '1'.  I'll put a comment to explain this.

File: esp/smc/SMCLib/TpWrapper.cpp (hunk 5 above)
Comment (shamser) [reply esp/smc/SMCLib/TpWrapper.cpp:498]: I've removed this line.  (Removal didn't cause any problems).

**Review by jakesmith — APPROVED** [review]
@shamser - looks good, please squash.

**Review by jakesmith — APPROVED** [review]
Looks good to merge.

**Pull request merged**